File: docs/examples/scala-lsp/shapes.scala
package demo

// Domain types used by main.scala.  The cross-file LSP actions live here.

trait Shape:                                   // Alt-Q R -> all references;
                                               // Alt-Q J on `Shape` -> SUBTYPES
                                               //   (Circle / Rectangle / Triangle)
  def area: Double                             // Alt-Q I on `area`  -> implementations
  def name: String                             // Alt-Q U on `name`  -> uses in this file

case class Circle(radius: Double) extends Shape:   // Alt-Q D from main.scala lands here;
                                                   // Alt-Q K on `Circle` -> SUPERTYPES (Shape)
  def area: Double = math.Pi * radius * radius     // Alt-Q H on `Pi` -> math.Pi: Double
  def name: String = "circle"

case class Rectangle(width: Double, height: Double) extends Shape:
  def area: Double = width * height
  def name: String = "rectangle"

case class Triangle(base: Double, height: Double) extends Shape:
  def area: Double = 0.5 * base * height
  def name: String = "triangle"

enum Color:                                    // Alt-Q W: type "Color" to jump here
  case Red, Green, Blue

case class Paint(shape: Shape, color: Color)

File: docs/examples/go-lsp/shapes.go
package main

// Domain types used by main.go.  The cross-file LSP actions live here.

// Shape is the interface the demo types satisfy.
// Alt-Q R on `Shape` -> all references;
// Alt-Q J on `Shape` -> SUBTYPES (the implementers: Circle / Rectangle / Triangle)
type Shape interface {
	Area() float64 // Alt-Q I on `Area` -> implementations
	Name() string  // Alt-Q U on `Name` -> uses in this file
}

// Alt-Q D from main.go lands here;
// Alt-Q K on `Circle` -> SUPERTYPES (the Shape interface it implements)
type Circle struct{ Radius float64 }

func (c Circle) Area() float64 { return 3.14159265358979 * c.Radius * c.Radius } // Alt-Q H on Area
func (c Circle) Name() string  { return "circle" }

type Rectangle struct{ Width, Height float64 }

func (r Rectangle) Area() float64 { return r.Width * r.Height }
func (r Rectangle) Name() string  { return "rectangle" }

type Triangle struct{ Base, Height float64 }

func (t Triangle) Area() float64 { return 0.5 * t.Base * t.Height }
func (t Triangle) Name() string  { return "triangle" }
